rkt/config/config.go

```diff
@@ -0,0 +1,276 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+)
+
+// Headerer is an interface for getting additional HTTP headers to use
+// when downloading data (images, signatures).
+type Headerer interface {
+	Header() http.Header
+}
+
+// Config is a single place where configuration for rkt frontend needs
+// resides.
+type Config struct {
+	AuthPerHost map[string]Headerer
+}
+
+type configParser interface {
+	parse(config *Config, raw []byte) error
+}
+
+var (
+	// configSubDirs is a map saying what kinds of configuration
+	// (values) are acceptable in a config subdirectory (key)
+	configSubDirs  = make(map[string][]string)
+	parsersForKind = make(map[string]map[string]configParser)
+)
+
+// Default paths for vendor and custom configuration
+const (
+	DefaultVendorPath = "/usr/lib/rkt"
+	DefaultCustomPath = "/etc/rkt"
+)
+
+func addParser(kind, version string, parser configParser) {
+	if len(kind) == 0 {
+		panic("empty kind string when registering a config parser")
+	}
+	if len(version) == 0 {
+		panic("empty version string when registering a config parser")
+	}
+	if parser == nil {
+		panic("trying to register a nil parser")
+	}
+	if _, err := getParser(kind, version); err == nil {
+		panic(fmt.Sprintf("A parser for kind %q and version %q already exist", kind, version))
+	}
+	if _, ok := parsersForKind[kind]; !ok {
+		parsersForKind[kind] = make(map[string]configParser)
+	}
+	parsersForKind[kind][version] = parser
+}
+
+func registerSubDir(dir string, kinds []string) {
+	if len(dir) == 0 {
+		panic("trying to register empty config subdirectory")
+	}
+	if len(kinds) == 0 {
+		panic("kinds array cannot be empty when registering config subdir")
+	}
+	allKinds := toArray(toSet(append(configSubDirs[dir], kinds...)))
+	sort.Strings(allKinds)
```

Comment:
redundant with 101 fwiw, could just define that func as returning sorted results (or really inline all that here since it isn't used elsewhere yet)